### PR TITLE
fix(mcp): expose concrete Codex input schemas

### LINKED ISSUE #386: bug(codex): purpose-review item schema collapses to Array<unknown>
## Why

ProjectAtlas v0.4.0 emits standards-valid typed MCP input schemas, but the supported Codex tool-schema bridge does not reliably retain local `$defs` and `$ref` members. Nested inputs such as `atlas_purpose_review.items` can therefore appear as `Array<unknown>` even though runtime deserialization remains strict.

## What Changes

- Audit the complete advertised MCP input inventory and keep every current Codex-facing schema self-contained.
- Inline the three affected small nested types at their existing Schemars ownership boundary without changing Serde or runtime request types.
- Keep `atlas_purpose_review.items` concrete with required `path` and optional purpose and conditional-write fields.
- Add packaged raw-stdio coverage that rejects local schema references, verifies the purpose-review shape, and confirms `tools/list` creates no project index state.
- Preserve raw MCP rejection, conditional-write admission, wrong-root, missing-index, and no-implicit-mutation behavior.
- Treat general Codex support for standards-valid local references as an upstream host dependency; ProjectAtlas supplies only the narrow compatibility representation it owns.

## Capabilities

- MCP tool-schema compatibility
- Typed purpose-review requests
- Codex plugin integration
- Raw stdio contract validation

## Architecture Diagrams

- [ProjectAtlas architecture views](https://github.com/styler-ai/ProjectAtlas/blob/dev/docs/projectatlas-3-architecture.md#architecture-views)

## Release Scope

Bugfix-only v0.4.1 work in the existing CLI/MCP schema annotations and packaged MCP contract tests. No database schema, storage behavior, CLI contract, dependency, tool inventory, or broad MCP redesign is included.

## Non-Goals

- Replacing typed request structs with untyped JSON values.
- Building a generic JSON Schema resolver or rewriting framework.
- Weakening runtime validation in reliance on host-side validation.
- Folding the broader tool-surface redesign from #310 into this bug.
- Claiming that local ProjectAtlas tests fix or prove Codex's general schema renderer.

## Pre-Mortem

Likely failure modes:

- Only purpose review is repaired while another advertised tool retains a bridge-sensitive local reference.
- A generic schema walker duplicates Schemars and creates a second schema authority.
- The exposed item object loses required `path` or makes optional conditional-write fields required.
- Host-readable schemas mask weaker runtime admission or allow an implicit database mutation.
- Raw stdio shape evidence is overstated as proof of Codex's upstream renderer.
- The packaged tool inventory changes without updating its compatibility contract.

Mitigations:

- [x] Reproduce the raw boundary, audit the complete inventory, and inline the exact affected typed schemas. (OpenSpec tasks: 1.1, 1.2, 2.1)
- [x] Gate every advertised input and verify the concrete purpose-review required and optional fields. (OpenSpec tasks: 2.2, 3.1)
- [x] Preserve typed runtime rejection, conditional writes, root isolation, missing-index behavior, and no mutation. (OpenSpec tasks: 2.3, 3.1)
- [x] Run the proportionate local Rust, OpenSpec, issue-map, and diff-quality gates. (OpenSpec tasks: 3.2)
- [x] Synchronize the hosted checklist and reconcile packaged-Codex and live review behavior. (OpenSpec tasks: 3.3)

## OpenSpec Tasks

OpenSpec change: `expose-concrete-codex-tool-schemas`

### 1. Reproduction and Contract

- [x] 1.1 Reproduce the v0.4.0 raw-stdio local-reference schema and identify the Codex bridge compatibility boundary.
- [x] 1.2 Audit all advertised MCP inputs for local `$defs` and `$ref` use and record the complete affected set.

### 2. Concrete MCP Schemas

- [x] 2.1 Inline the affected schemas at their nested typed-schema boundary without changing Serde or runtime request types.
- [x] 2.2 Add packaged raw-stdio coverage that rejects any advertised local input reference and verifies concrete purpose-review required and optional fields.
- [x] 2.3 Preserve and verify runtime rejection, conditional-write admission, wrong-root, missing-index, and no-implicit-mutation behavior.

### 3. Verification

- [x] 3.1 Run focused MCP schema and purpose-review tests, including the missing-`path` negative case and compatibility inventory.
- [x] 3.2 Run formatting, workspace check, Clippy, rustdoc, OpenSpec, issue-map/checklist diagnostics, and diff-quality gates proportionate to the change.
- [x] 3.3 Synchronize the GitHub checklist and obtain final integrated, packaged-Codex/hosted, and live review proof.

### PR DESCRIPTION
**Pull request by styler-ai**
## Summary

- emit self-contained input schemas for every currently affected MCP nested type
- keep purpose-review path required and its optional fields concrete without changing typed runtime admission
- audit all advertised schemas through the packaged raw-stdio contract and preserve no-mutation behavior

Closes #386

## Verification

- focused raw MCP tools/list and missing-path regressions
- full ordinary workspace, CLI E2E, installer, parser, documentation, dependency, OpenSpec, and IssueOps gates
- long benchmark campaigns remain manual-only and were not run

### CODE REVIEW

**Comment by styler-ai** [comment]
@codex review

**Comment by styler-ai** [comment]
@codex review

File: openspec/changes/expose-concrete-codex-tool-schemas/tasks.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+## 1. Reproduction and Contract
+
+- [x] 1.1 Reproduce the v0.4.0 raw-stdio local-reference schema and identify the Codex bridge compatibility boundary.
+- [x] 1.2 Audit all advertised MCP inputs for local `$defs` and `$ref` use and record the complete affected set.
+
+## 2. Concrete MCP Schemas
+
+- [x] 2.1 Inline the affected schemas at their nested typed-schema boundary without changing Serde or runtime request types.
+- [x] 2.2 Add packaged raw-stdio coverage that rejects any advertised local input reference and verifies concrete purpose-review required and optional fields.
+- [x] 2.3 Preserve and verify runtime rejection, conditional-write admission, wrong-root, missing-index, and no-implicit-mutation behavior.
```
[the remark this answers is not in the document]
Comment (styler-ai) [reply openspec/changes/expose-concrete-codex-tool-schemas/tasks.md:10]: Added the missing real stdio boundary cases in the existing packaged MCP contract test: incomplete conditional identity errors, fully stale identity returns a typed stale conflict, a non-directory root errors, and a missing-index root errors without creating project state. Each case preserves the logical database snapshot; the focused packaged test passes.

**Comment by styler-ai** [comment]
@codex review

File: crates/projectatlas-cli/tests/e2e.rs [hunk 2]
```diff
@@ -19129,6 +19229,148 @@ fn mcp_tools_by_name(tools: &[Value]) -> Result<BTreeMap<&str, &Value>, Box<dyn
...
+/// Reject bridge-sensitive local definitions or references anywhere in an input schema.
+fn assert_no_local_schema_references(path: &str, value: &Value) -> Result<(), Box<dyn Error>> {
+    match value {
+        Value::Object(object) => {
+            for (key, child) in object {
+                if key == "$defs"
+                    || key == "$ref"
+                        && child
+                            .as_str()
+                            .is_some_and(|reference| reference.starts_with("#/$defs/"))
```
[the remark this answers is not in the document]
Comment (styler-ai) [reply crates/projectatlas-cli/tests/e2e.rs]: Fixed. The recursive inventory audit now rejects every $ref plus both $defs and legacy definitions containers, so local fragments, external references, and definition-backed schemas cannot pass as self-contained. A focused fixture covers each form, and the real packaged tools-list schema audit remains green.

**Comment by styler-ai** [comment]
@codex review

**Comment by styler-ai** [comment]
@codex review
